tests/e2e/fullsync_test_for_block_volume.go

```diff
@@ -901,3 +918,219 @@ var _ bool = ginkgo.Describe("full-sync-test", func() {
+func waitAndVerifyCnsVolumeMetadata4GCVol(volHandle string, svcPVCName string, pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume, pod *v1.Pod) error {
+
+	waitErr := wait.PollImmediate(poll*5, pollTimeoutShort, func() (bool, error) {
+		matches := verifyCnsVolumeMetadata4GCVol(volHandle, svcPVCName, pvc, pv, pod)
+		return matches, nil
+	})
+	return waitErr
+}
+
+//verifyCnsVolumeMetadata4GCVol verifies cns volume metadata for a GC volume
+// if gcPvc, gcPv or pod are nil we skip verification for them altogether and wont check if they are absent in CNS entry
+func verifyCnsVolumeMetadata4GCVol(volumeID string, svcPVCName string, gcPvc *v1.PersistentVolumeClaim,
+	gcPv *v1.PersistentVolume, pod *v1.Pod) bool {
+
+	cnsQueryResult, err := e2eVSphere.queryCNSVolumeWithResult(volumeID)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	if cnsQueryResult.Volumes == nil || len(cnsQueryResult.Volumes) == 0 {
+		framework.Logf("CNS volume query yielded no results for volume id: " + volumeID)
+		return false
+	}
+	cnsVolume := cnsQueryResult.Volumes[0]
+	podEntryFound := false
+	verifyPvEntry := false
+	verifyPvcEntry := false
+	verifyPodEntry := false
+	gcPVCVerified := false
+	gcPVVerified := false
+	svcPVCVerified := false
+	svcPVVerified := false
+	svcPVC := getPVCFromSupervisorCluster(svcPVCName)
+	svcPV := getPvFromSupervisorCluster(svcPVCName)
+	if gcPvc != nil {
+		verifyPvcEntry = true
+	} else {
+		gcPVCVerified = true
+	}
+	if gcPv != nil {
+		verifyPvEntry = true
+	} else {
+		gcPVVerified = true
+	}
+	if pod != nil {
+		verifyPodEntry = true
+	}
+	framework.Logf("Found CNS volume with id %v\n"+spew.Sdump(cnsVolume), volumeID)
+	gomega.Expect(cnsVolume.Metadata).NotTo(gomega.BeNil())
+	for _, entity := range cnsVolume.Metadata.EntityMetadata {
+		var pvc *v1.PersistentVolumeClaim
+		var pv *v1.PersistentVolume
+		entityMetadata := entity.(*cnstypes.CnsKubernetesEntityMetadata)
+		if entityMetadata.EntityType == string(cnstypes.CnsKubernetesEntityTypePVC) {
+			verifySvcPvc := false
+			verifySvcPv := false
+			if entityMetadata.EntityName == svcPVCName {
+				pvc = svcPVC
+				pv = svcPV
+				verifySvcPvc = true
+				verifySvcPv = true
+			} else {
+				pvc = gcPvc
+				pv = gcPv
+			}
+			if verifyPvcEntry || verifySvcPvc {
+				if entityMetadata.EntityName != pvc.Name {
+					framework.Logf("PVC name '%v' does not match PVC name in metadata '%v', for volume id %v",
+						pvc.Name, entityMetadata.EntityName, volumeID)
+					break
+				}
+				if verifyPvEntry || verifySvcPv {
+					if entityMetadata.ReferredEntity == nil {
+						framework.Logf("Missing ReferredEntity in PVC entry for volume id %v", volumeID)
+						break
+					}
+					if entityMetadata.ReferredEntity[0].EntityName != pv.Name {
+						framework.Logf("PV name '%v' in referred entity does not match PV name '%v', "+
+							"in PVC metadata for volume id %v", entityMetadata.ReferredEntity[0].EntityName,
+							pv.Name, volumeID)
```

Comment:
entityMetadata.ReferredEntity[0].EntityName, pv.Name,
should be
pv.Name, entityMetadata.ReferredEntity[0].EntityName,

Reply:
looks fine to me.